v3

mdwhatcott opened the pull request:
The sqldb module has had a very 'enterprise' feel to it, but it's not really doing that much. This PR proposes an implementation that retains the behavior of previous versions but does so with less than 100 lines of code. It also adds inherent support for prepared statements, as
well as `QueryRowContext`.

See this commit for the new implementation at a glance: https://github.com/smarty/sqldb/pull/2/commits/72958119d3c2db94a8c2c5252e7e5aee8881f0db

mdwhatcott committed:
Reset

mdwhatcott committed:
v3

Implement the spirit of previous versions in a single file with just a
few simple functions. Adds inherent support for prepared statements, as
well as QueryRowContext.

It is very plausible that this module, which is now less than 100 lines
of code, is a good candidate for copy-paste installation.

mdwhatcott committed:
Use SplitSeq

mdwhatcott committed:
Named return values

bendorton commented on sqldb.go:
```diff
@@ -0,0 +1,91 @@
+		if err != nil {
+			return 0, err // already normalized
+		}
+		count += rows
```
Is it helpful to sum the count of affected rows if multiple statements are being executed? Would it be more helpful to return a slice of counts or count of successful statement executions?

joliver replied:
The need to split multiple statements into multiple execution statements is actually a limitation of the MySQL driver specifically. Other drivers (pgsql) can handle multiple statements. The split with the subsequent aggregation of rows affected is masking the fact that it's multiple, separate statements being executed.

The count as a single integer is important for optimistic concurrency, e.g. 27 rows in total were affected and we expected 30 rows affected therefore something must have changed underneath us.

mdwhatcott commented:
@kierstenSmarty - I don't really have immediate plans, but perhaps we should be considering this for some of our more active paths.

mdwhatcott committed:
Integration test suite (in separate module). ⤵️

..lest we pollute the main module's dependency listing.

This suite utilizes an in-memory sqlite database to avoid any need for a
running database server.

mdwhatcott committed:
Return rows.Err as final result of BindAll func.

mdwhatcott committed:
Export error normalization func.

mdwhatcott committed:
Document NormalizeErr func.

mdwhatcott committed:
Function to bind an [optional] row.

mdwhatcott committed:
Expand API with a DB that can execute 'Scripts' and 'Queries'. ⤵️

This DB leverages prepared statements when configured thresholds are
encountered.

The DB type is the approach I'd like to use for new code.
Existing mappers can be more easily transitioned to the package-level
functions that were implemented in previous commits on this v3 branch.

Perhaps those functions will be moved to a separate package...

mdwhatcott committed:
Context cancellation needs no stack trace.

mdwhatcott committed:
Remove duplication.

mdwhatcott committed:
Committing to (what is now called) the Handle interface.

No more package-level/static functions.

mdwhatcott committed:
Some notes.

mdwhatcott committed:
Define RowsAffected interface as optional hook for Script types.

mdwhatcott committed:
Tweak wording.

mdwhatcott committed:
Intentionally defining contractual interfaces to be exported. ⤵️

I'm allowing these contractual interfaces to be exported so they can be
conveniently referenced by calling code. This decision was made after
careful consideration of intended implementation patterns as I actually
coded against the new API, which helped determine that the ability to
reference these type definitions provides significant value.

mdwhatcott committed:
Lower threshold to trigger use of prepared statements.

mdwhatcott committed:
Finesse the docs.

mdwhatcott committed:
Base types, for convenience.

mdwhatcott committed:
Store checksums of statements (smaller) and cap how many we will track.

mdwhatcott committed:
go work sync

mdwhatcott committed:
Include adapter code to ease migration from v2 to v3.

mdwhatcott committed:
Variadic methods.

mdwhatcott committed:
Test interleaving of statements and parameters.
